### DISCUSSION TITLE
OrdScript - an Ordinals native scripting language for inscriptions

### DISCUSSION THREAD

**Discussion by huuep** [ordinals/ord#2217] [Ideas]
Instead of using JS, or some other language to program in an inscription, OrdScript can be an Ordinals native scripting language that is optimized for Ordinals. How it could work is that it can be automatically run in browsers by first loading a recursive inscription that is the run time compiler for OrdScript. Here's why we would want OrdScript:

- the language is designed to be very effecient in program size (like GolfScript) to save Bitcoin block space
- the language has Bitcoin and Ordinal primitives and functions built in
- the language has some safeguards/limitations to prevent malicious programming
- the language allows for easy coding using recursive inscription calls, library calls to other inscriptions, and working with compression
- the language allows for easy coding using reinscription, especially in terms of using it to store/reference state changes

**Comment by tansanDOTeth** [comment]
> the language has some safeguards/limitations to prevent malicious programming

Unless OrdScript is mandatory, it's easy to escape these safeguards by just using JS. If it's not mandatory, wouldn't a JS SDK be better?

I would argue OrdScript is a nerf unless you can inherit the JS ecosystem which is so rich and vast. 

If you guys do make a SDK, I would love for it to have access to data like current inscription SAT.

**Comment by bodily11** [comment]
I think with a JS SDK you get most of what you are looking for:
1. SDK is efficient because you only have to inscribe the SDK once and it can easily be loaded into other inscriptions
2. SDK has bitcoin and ordinals primitives built-in
3. I think safeguards and limitations need to be built into explorers, wallets, and marketplaces anyway, so a client side SDK could serve this purpose
4. Recursion / compression / calling other libraries could all be built into SDK
5. Reinscription / sat references for storage and referencing other inscriptions could be built-in to SDK

To me, taking advantage of the network effects, libraries, and native browser support of JS makes any additional scripting language a hard sell. Efficiency (every byte matters) is an interesting argument, but I'm not sure its enough.
